Sprint 4/step 1

dqrest opened the pull request.

dqrest committed:
jsx ---> tsx

dqrest committed:
modal: jsx ---> tsx

dqrest committed:
- dirty code

dqrest committed:
IngredientType: jsx ---> tsx

dqrest committed:
const ---> enum

dqrest committed:
big step: jsx ---> tsx

dqrest committed:
the secong big step: jsx ---> tsx for pages

dqrest committed:
order-details, burger-constructor: jsx ---> tsx

dqrest committed:
profile-menu: jsx ---> tsx

dqrest committed:
app-header: jsx ---> tsx

dqrest committed:
error-boundary: jsx ---> tsx

dqrest committed:
protected-route-element: jsx ---> tsx

dqrest committed:
app: jsx ---> tsx

gennady-bars commented on src/components/protected-route-element/protected-route-element.tsx:
```diff
@@ -1,10 +1,11 @@
 
-export default function ProtectedRouteElement({ user, element }) {
+
+const ProtectedRouteElement: FC<TProtectedRouteElement> = ({ user, element }): ReactElement => {
```
**Можно лучше**

Если интересно, то можно сделать вот так универсальный компонент `ProtectedRoute` с переадресацией

```jsx


export default function ProtectedRoute({ children, anonymous = false }) {
  const isLoggedIn = useSelector((store) => store.user.isLoggedIn);

  const location = useLocation();
  const from = location.state?.from || '/';
  // Если разрешен неавторизованный доступ, а пользователь авторизован...
  if (anonymous && isLoggedIn) {
    // ...то отправляем его на предыдущую страницу
    return <Navigate to={ from } />;
  }

  // Если требуется авторизация, а пользователь не авторизован...
  if (!anonymous && !isLoggedIn) {
    // ...то отправляем его на страницу логин
    return <Navigate to="/login" state={{ from: location}}/>;
  }

  // Если все ок, то рендерим внутреннее содержимое
  return children;
}

```

dqrest replied:
Сделано:
1)  добавил компонент app.tsx
2) переделан checkResponse.tsx

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

К сожалению, работу придется отклонить без потери итерации, так как проект не запускается. Скрин ошибки  https://disk.yandex.ru/i/jOw97n6X55CiQg 


* `check-response.js` нужно было тоже типизировать
*  Все `PropTypes` теперь нужно удалить из проекта. Они остались в коде

dqrest committed:
clear code

dqrest committed:
app: jsx ---> tsx

dqrest committed:
fix bug

dqrest committed:
fix bug

Commit:
remove unused code

gennady-bars commented on src/utils/auth-api.ts:
```diff
@@ -0,0 +1,161 @@
+    , TResetPasswordFormData
+} from '../shared/types/auth-types';
+
+const NORMA_API = "https://norma.nomoreparties.space/api";
```
`const NORMA_API = "https://norma.nomoreparties.space/api"; `  нужно объявить только 1 раз в `api.ts`. Нигде больше не нужно дублировать базовый урл, чтобы легко его можно было бы исправить

dqrest replied:
сделано

gennady-bars commented on src/utils/burger-api.ts:
```diff
@@ -0,0 +1,10 @@
+import { TResponseBody } from './api';
+import { TBurgerIngredientsItemDto } from '../shared/dtos/burger-ingredients-item-dto';
+
+const NORMA_API = "https://norma.nomoreparties.space/api";
```
`const NORMA_API = "https://norma.nomoreparties.space/api"; `  нужно объявить только 1 раз в `api.ts`. Нигде больше не нужно дублировать базовый урл, чтобы легко его можно было бы исправить

dqrest replied:
сделано

gennady-bars commented on src/utils/order-api.ts:
```diff
@@ -0,0 +1,19 @@
+import { TResponseBody } from './api';
+import { TBurgerOrderItemDto } from '../shared/dtos/burger-order-item-dto';
+
+const NORMA_API = "https://norma.nomoreparties.space/api";
```
`const NORMA_API = "https://norma.nomoreparties.space/api"; `  нужно объявить только 1 раз в `api.ts`. Нигде больше не нужно дублировать базовый урл, чтобы легко его можно было бы исправить

dqrest replied:
сделано

gennady-bars commented on src/services/actions/burger-incredients.ts:
```diff
@@ -0,0 +1,91 @@
+        dispatch(getIngredientsAction());
+        getIngredientsRequest()
+            .then(res => {
+                if (res && res.success) {
```
`                if (res && res.success) {
`

Это дублируется везде


**Можно лучше**


Если интересно, то можно создать **универсальную** функцию запроса `request` , которая будет содержать в себе уже **базовый урл**, проверку на **ok** и проверку на `success`, чтобы не дублировать их в каждом запросе.

```js

// 1 раз объявляем базовый урл
export const BASE_URL = "https://norma.nomoreparties.space/api/";

// создаем функцию проверки ответа на `ok`
const checkResponse = (res) => {
  if (res.ok) {
    return res.json();
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ошибка ${res.status}`);
};

// создаем функцию проверки на `success`
const checkSuccess = (res) => {
  if (res && res.success) {
    return res;
  }
  // не забываем выкидывать ошибку, чтобы она попала в `catch`
  return Promise.reject(`Ответ не success: ${res}`);
};

// создаем универсальную фукнцию запроса с проверкой ответа и `success`
// В вызов приходит `endpoint`(часть урла, которая идет после базового) и опции
const request = (endpoint, options) => {
  // а также в ней базовый урл сразу прописывается, чтобы не дублировать в каждом запросе
  return fetch(`${BASE_URL}${endpoint}`, options)
    .then(checkResponse)
    .then(checkSuccess);
};

```


И теперь не нужно будет дублировать все это в каждом запросе. Нужно просто заменить в запросах все `fetch` на `request`, удалив лишние проверки и базовый урл. Все остальное останется без изменений

Код станет намного чище. Будет мелкий аналог `axios`

Пример запроса ингредиентов

```js
const getIngredients =  () => request("ingredients");

```

В `get`-запросах даже не нужно указывать 2й параметр. Он получится `undefined`, что без проблем для `fetch` (который внутри).

dqrest replied:
сделано

gennady-bars commented on src/pages/login/index.tsx:
```diff
@@ -9,19 +9,21 @@ import {
     const { signIn, request, failed, message } = useProvideAuth();
-    const [formData, setFormData] = useState({
+
+    const [formData, setFormData] = useState<TLoginFormData>({
```
**Можно лучше**

Можно сделать универсальный **кастомный** хук для контроля любого количества инпутов в любых формах:

```js
export function useForm(inputValues={}) {
  const [values, setValues] = useState(inputValues);

  const handleChange = (event) => {
    const {value, name} = event.target;
    setValues({...values, [name]: value});
  };
  return {values, handleChange, setValues};
}

```

Этот код помещают в отдельный файл `useForm.js` в папке `hooks` и импортируют функцию туда, где нужно контролировать инпуты

И Вам не нужно будет теперь вручную создавать функции обработки инпутов и т д. Все будет в одной строчке кода:

```jsx
  const {values, handleChange, setValues} = useForm({});

```

gennady-bars commented on src/pages/reset-password/index.tsx:
```diff
@@ -14,22 +16,21 @@ import {
+    });
 
     const errorMessage = (
         <span className='text text_type_main-default' style={{ color: 'red' }}>
```
Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Это затрудняет управление стилями. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили.


и в других местах тоже нужно исправить по коду


Скрин https://disk.yandex.ru/i/wEd56B_Zt8JHUw 28 раз используете в коде это

gennady-bars commented on src/components/app/app.tsx:
```diff
@@ -61,25 +65,28 @@ function LoadingApp() {
-ProtectedApp.propTypes = {
-    user: userDto
-}
+
```
` <AppHeader />` нужно поместить только в `App`, чтобы любая страница была с этим хэдером. Не нужно дублировать его в каждой странице

gennady-bars commented on src/components/protected-route-element/protected-route-element.tsx:
```diff
@@ -46,13 +47,12 @@ export default function ProtectedRouteElement({ user, element }) {
         return <Navigate to={fb || '/'} replace />;
     }
```
**Можно лучше**

много хардкода внутри `ProtectedRouteElement`

```js

  if (!user) {

        if (pathname.includes('/login')
            || pathname.includes('/forgot-password')
            || pathname.includes('/register'))
            return element;

        if (pathname.includes('/reset-password')
            && location?.state?.access
            && location?.state?.email
            && location?.state?.email?.length > 0) {
            return element;
        }

        if (pathname.includes('/profile')
            || pathname.includes('/profile/user')
            || pathname.includes('/profile/orders')) {
            fb = pathname;
        }
        const fbp = fb
            ? `?fallback=${fb}`
            : '';

        return <Navigate to={`/login${fbp}`} replace />;
    }

```

gennady-bars replied:
**Можно лучше**

Если интересно, то можно сделать вот так универсальный компонент `ProtectedRoute` с переадресацией

```jsx


export default function ProtectedRoute({ children, anonymous = false }) {
  const isLoggedIn = useSelector((store) => store.user.isLoggedIn);

  const location = useLocation();
  const from = location.state?.from || '/';
  // Если разрешен неавторизованный доступ, а пользователь авторизован...
  if (anonymous && isLoggedIn) {
    // ...то отправляем его на предыдущую страницу
    return <Navigate to={ from } />;
  }

  // Если требуется авторизация, а пользователь не авторизован...
  if (!anonymous && !isLoggedIn) {
    // ...то отправляем его на страницу логин
    return <Navigate to="/login" state={{ from: location}}/>;
  }

  // Если все ок, то рендерим внутреннее содержимое
  return children;
}

```

gennady-bars commented on src/components/modal/modal.tsx:
```diff
@@ -0,0 +1,78 @@
+        // buttonclick не на modal ---> скрыть Modal        
+        const target = e?.target as HTMLElement;
+        if (overlayRef?.current?.className === target?.className)
+            props.setVisible(false);
```
` props.setVisible(false);` вынесено в `closeIconClick`. Нужно ее назвать нейтрально `closePopup` и использовать в оверлее, `Escape` и крестике

dqrest replied:
сделано

gennady-bars commented on src/components/modal/modal.tsx:
```diff
@@ -0,0 +1,78 @@
+    }     
+
+    const closeIconClick: MouseEventHandler = (e) => {        
+        props.setVisible(false);
```
` props.setVisible(false);` вынесено в `closeIconClick`. Нужно ее назвать нейтрально `closePopup` и использовать в оверлее, `Escape` и крестике

dqrest replied:
сделано

gennady-bars commented on src/components/modal/modal.tsx:
```diff
@@ -0,0 +1,78 @@
+    useEffect(() => {
+        // Нажали на Escape ---> скрыть Modal
+        const handleEscapeKey = (event: KeyboardEvent) =>
+            event?.code === 'Escape' && props?.setVisible(false);
```
` props.setVisible(false);` вынесено в `closeIconClick`. Нужно ее назвать нейтрально `closePopup` и использовать в оверлее, `Escape` и крестике

gennady-bars commented on src/components/burger-ingredients/burger-ingredients.tsx:
```diff
@@ -9,46 +9,54 @@ import {
     const dispatch = useDispatch();
     useEffect(() => {
         // load all burder ingredients
         dispatch(getIngredients());
```
`dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.

dqrest replied:
сделано

gennady-bars commented on src/components/burger-constructor/burger-constructor.tsx:
```diff
@@ -128,7 +135,7 @@ export default function BurgerConstructor() {
-                            ing.type !== IngredientType.Bun &&
+                            ing.type !== TIngredient.Bun &&
                             <BurgerDraggableConstructorItem
                                 key={`${ind}_${ing._id}_wrapper`}
```
```js

                                key={`${ind}_${ing._id}_wrapper`}


```



Для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время **сброса в зону** `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND.


 А в `key` будет уже просто **key={ item.uniqueId }**

gennady-bars replied:
Нужно генерировать uuid при создании экшена (желательно, в **`action creator`**). И созданный uuid должен приходить в экшене в `payload` или как отдельное поле, или сразу как часть объекта ингредиента. Посоветуйтесь с наставником, пожалуйста.
```js

// вот это - `action creator`
export const addIngridient = (item) => {
    return {
        type: ADD_INGRIDIENT,
        payload: {
            ...item, // используем `spread`, чтобы поменять ссылку на объект. Таким образом `redux` обновит его в хранилище
           uniqueId: uuid4()  // и добавляем в объект новое поле, которое потом будет использовано в `key`
        }
    }
}


```

dqrest replied:
сделано

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Посмотрите на **гитхабе** все комментарии к коду (нужно прокрутить вниз страницу там)

Работа проделана огромная:



* Хорошо типизировали компоненты


но есть некоторые недочеты:




* `const NORMA_API = "https://norma.nomoreparties.space/api"; `  нужно объявить только 1 раз в `api.ts`. Нигде больше не нужно дублировать базовый урл, чтобы легко его можно было бы исправить
* Если **перезагрузить** страницу с открытым попапом ингредиента, то этот открытый попап опять должен появляться (или страница с деталями). А сейчас ничего не отображается после перезагрузки. 
* Инлайн-стилей `style={{`не должно быть в проекте, если они не меняются **динамически** в коде компонента. Это затрудняет управление стилями. Все стили описываются в модульных `.module.css` файлах, а в коде компонентов используются модульные `css`-стили. Скрин https://disk.yandex.ru/i/wEd56B_Zt8JHUw 28 раз используете в коде это
* ` <AppHeader />` нужно поместить только в `App`, чтобы любая страница была с этим хэдером. Не нужно дублировать его в каждой странице
* `dispatch(getIngredients())` достаточно 1 раз сделать в `App`, чтобы все страницы получили ингредиенты при загрузке сайта. Не нужно дублировать запрос в других компонентах.
* Для атрибута `key` в конструкторе нужно использовать для ингредиента уникальный `uuid` (Вам нужно его генерировать во время **сброса в зону** `DnD`). Это можно делать с помощью библиотеки `uuid` https://www.npmjs.com/package/uuid , или Вам должны были показать, как это сделать, иначе некорректно может работать DND. А в `key` будет уже просто **key={ item.uniqueId }**



Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

Commit:
fix bugs:
1) show modal with ingredients after refreshing the browser
2) get ingredients in app.tsx

Commit:
mover AppHeader, ErrorBoundary from home to app.tsx

Commit:
move NORMA_API to api.ts

Commit:
move styles to className

Commit:
move style items from styles to className

Commit:
move style items from styles to className

Commit:
closeIconClick ---> closePopup

Commit:
use uuid for map ingredients in burgerconstructor

Commit:
+ checkSuccess

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.
